subiquity/common/api/server.py

```diff
@@ -0,0 +1,172 @@
+        self.controller = controller
+        self.methname = methname
+
+    def __str__(self):
+        return f"{self.controller} must have method called {self.methname}"
+
+
+class SignatureMisatchError(BindError):
+    def __init__(self, methname, expected, actual):
+        self.methname = methname
+        self.expected = expected
+        self.actual = actual
+
+    def __str__(self):
+        return (f"implementation of {self.methname} has wrong signature, "
+                f"should be {self.expected} but is {self.actual}")
+
+
+def trim(text):
+    if text is None:
+        return ''
+    elif len(text) > 80:
+        return text[:77] + '...'
+    else:
+        return text
+
+
+def _make_handler(controller, definition, implementation, serializer):
+    def_sig = inspect.signature(definition)
+    def_ret_ann = def_sig.return_annotation
+    def_params = def_sig.parameters
+
+    impl_sig = inspect.signature(implementation)
+    impl_params = impl_sig.parameters
+
+    data_annotation = None
+    data_arg = None
+    query_args_anns = []
+
+    check_def_params = []
+
+    for param_name, param in def_params.items():
+        if param_name in ('request', 'context'):
+            raise Exception(
+                "api method {} cannot have parameter called request or "
+                "context".format(definition))
+        if getattr(param.annotation, '__origin__', None) is Payload:
+            data_arg = param_name
+            data_annotation = param.annotation.__args__[0]
+            check_def_params.append(param.replace(annotation=data_annotation))
+        else:
+            query_args_anns.append(
+                (param_name, param.annotation, param.default))
+            check_def_params.append(param)
+
+    check_impl_params = [
+        p for p in impl_params.values()
+        if p.name not in ('context', 'request')
+        ]
+    check_impl_sig = impl_sig.replace(parameters=check_impl_params)
+
+    check_def_sig = def_sig.replace(parameters=check_def_params)
+
+    if check_impl_sig != check_def_sig:
+        raise SignatureMisatchError(
+            definition.__qualname__, check_def_sig, check_impl_sig)
+
+    async def handler(request):
+        context = controller.context.child(
+            implementation.__name__, trim(await request.text()))
+        with context:
+            context.set('request', request)
+            args = {}
+            if data_annotation is not None:
+                payload = json.loads(await request.text())
+                args[data_arg] = serializer.deserialize(
+                    data_annotation, payload)
+            for arg, ann, default in query_args_anns:
+                if arg in request.query:
+                    v = serializer.deserialize(
+                        ann, json.loads(request.query[arg]))
+                elif default != inspect._empty:
+                    v = default
+                else:
+                    1/0
```

Comment:
Why division by zero here? Is it just to throw an exception if we get here?

Reply:
Yeah "1/0" is my lazy way of writing "assert False" and I should have cleaned this up before sending it for review!

This is the case of failing to provide a required argument in the api call, something the client side of this won't let you do but it should probably still be handled a little bit gracefully. I'll add a test case and make this more sensible.